File: logb-py/rule.py
from pattern import Pattern, Variable

class Rule:
    """ Inspired by prolog, this class represents a "ForAll implication". 
    
        To be precise:
    
        ForAll %X: A(%X) => B(%X),
        where %X can be any number of arguments, 
        and A() and B() can be any statement.
    """

    def __init__(self):
        self.variables = []
        self.dependencies = []
        self.conclusion = None

    def add_variable(self, name):
        rval = Variable(name)
        self.variables.append(rval)
        return rval

    def add_dependency(self, dependency):
        """ Add a new dependency for this rule. """
        self.dependencies.append(dependency)

    def set_conclusion(self, conclusion):
        """ Add a new conclusion for this rule. """
        self.conclusion = conclusion

    def to_statement(self):
        """ Creates a regular statement, consisting of ForAll and implication, for this rule. """

        # TODO
        
    def conclusion_pattern(self):
        """ Checks whether the given statement matches this rule's conclusion(head) and if so, returns the substitutions """
        
        # Build a new pattern with our conclusion
        pattern_to_check = Pattern()
        pattern_to_check.root = self.conclusion
        pattern_to_check.variables = self.variables
        
        return pattern_to_check
    
    def get_dependencies(self, substitutions):
        """ Get the dependencies with all variables substituted through `substitutions`. """
        
        rval = []
        
        for dependency in self.dependencies:
            # For each dependency, we create a pattern from it, then substitute all variables.
            # Once finished, all variables should be gone(there should have been substitutions for them).
            # Since there no longer are any variables, we can remove the statement from its pattern,
            # and just return the statement itself.
            used_pattern = Pattern()
            used_pattern.root = dependency
            used_pattern.variables = self.variables
            
            used_pattern = used_pattern.substitute(substitutions)
            
            #if len(used_pattern.variables) > 0:
            #    # Not all variables have been substituted, oops!
            #    return None
            
            rval.append(used_pattern.root)
            
        return rval

File: logb-java/src/org/logb/core/Rule.java
package org.logb.core;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Inspired by prolog, this class represents a "ForAll implication". To be
 * precise:
 * 
 * <pre>
 * <code>
 *         ForAll %X: A(%X) => B(%X),
 *         where %X can be any number of arguments, 
 *         and A() and B() can be any statement.
 *    </code>
 * </pre>
 * 
 * <code>`B(%X)`</code> is called the head of the rule. <code>`A(%X)`</code> is
 * the tail.
 */
public class Rule {
	/**
	 * Tries to add a variable of the given name, ensures the name is unique.
	 * 
	 * If the name is not unique, a different name will be selected for you. The
	 * new name can be extracted from the return value by calling `getName()`.
	 * 
	 * @param name
	 *            The name to give the variable, null to have the name selected
	 *            for you.
	 * @return The variable that was added.
	 */
	public Variable addVariable(String name) {
		boolean nameUnique = false;
		while (!nameUnique) {
			nameUnique = true;
			for (Variable var : variables) {
				if (var.getName() == name) {
					nameUnique = false;
				}
			}

			if (!nameUnique) {
				// Append _ until the name is unique
				name = name + "_";
			}
		}

		Variable rval = new Variable(name);
		variables.add(rval);
		return rval;
	}

	/**
	 * Add a dependency. In the rule (A && B && C) => D, A, B and C are
	 * individual dependencies that can be added with this method.
	 * 
	 * @param dependency The statement to add as dependency for the rule
	 */
	public void addDependency(EntityStructureBase dependency) {
		dependencies.add(dependency);
	}

	/**
	 * Sets the conclusion(head) of the rule.
	 * 
	 * @param conclusion The conclusion to set.
	 */
	public void setConclusion(EntityStructureBase conclusion) {
		this.conclusion = conclusion;
	}

	/**
	 * @return The conclusion of this rule as pattern.
	 */
	public Pattern getConclusionPattern() {
		Pattern rval = new Pattern();
		for(Variable var : variables) {
			rval.addVariable(var.getName());
		}
		rval.setRoot(this.conclusion);
		return rval;
	}

	/**
	 * Return this rule's dependencies, with variables substituted through the
	 * given substitutions.
	 * 
	 * @param substitutions
	 *            Maps variable names to strings to substitute them with.
	 * @return The substituted list of statements.
	 */
	public List<Statement> getDependencies(
			Map<String, EntityStructureBase> substitutions) {
		List<Statement> rval = new ArrayList<Statement>();
		
		for(EntityStructureBase dependency: dependencies) {
			// In order to apply the substitutions to the statement,
			// we first pack the statement into a pattern with the
			// same variables as this rule, then call substitute
			// on the pattern, then extract the resulting statement.
			
			Pattern tmpPattern = new Pattern();
			tmpPattern.setRoot(dependency);
			tmpPattern = tmpPattern.substitute(substitutions);
			
			// We know it's a Statement because we just set the root
			// with pattern.setRoot, and substitute only changes variables,
			// not existing statements.
			rval.add((Statement) tmpPattern.getRoot());
		}
		return rval;
	}
	
	public Set<String> getVariableNames() {
		Set<String> rval = new HashSet<>();
		
		for(Variable variable : variables) {
			rval.add(variable.getName());
		}
		
		return rval;
	}

	private List<Variable> variables = new ArrayList<Variable>();
	private List<EntityStructureBase> dependencies = new ArrayList<EntityStructureBase>();
	private EntityStructureBase conclusion;
}